anonymous_helpline_chatbot/core/chat_bot_core.py

```diff
@@ -26,15 +24,51 @@ def __dir__(self) -> Set[str]:
-            dir(self._conversations_controller),
-            dir(self._invitations_controller)
+            dir(self._conversations_controller)
+            # Not `self._invitations_controller`, to not confuse users of `ChatBotCore` with invitation functions,
+            # which they actually shouldn't use
         )
 
     def __getattr__(self, item):
-        for controller in (self._users_controller, self._conversations_controller, self._invitations_controller):
+        # Not searching in `self._invitations_controller`, to not confuse users of `ChatBotCore` with invitation
+        # functions, which they actually shouldn't use
+        for controller in (self._users_controller, self._conversations_controller):
             if hasattr(controller, item):
                 return controller.__getattribute__(item)
 
         # If reached this point, then `item` is not defined in any of the controllers. Produce `AttributeError`
         # (by calling `object.__getattribute__`):
         return super().__getattribute__(item)
+
+    @contextmanager
+    def request_conversation_with_locking(self, client_chat_id: int) -> Generator[int, None, None]:
+        with self._conversations_controller.request_conversation_with_locking(client_chat_id) as res:
+            if res == 0:
+                self._invitations_controller.invite_to_client(client_chat_id)
+            yield res
+
+    @contextmanager
+    def begin_conversation_with_locking(self, client_chat_id: int, operator_chat_id: int) -> Generator[int, None, None]:
+        with self._conversations_controller.begin_conversation_with_locking(client_chat_id, operator_chat_id) as res:
+            if res == 0:
+                self._invitations_controller.clear_invitations_to_client(client_chat_id)
+                # Not clearing invitations to client `operator_chat_id` because there mustn't be any because `res == 0`
+                self._invitations_controller.clear_invitations_for_operator(operator_chat_id)
+                # In case user `client_chat_id` is an operator:
+                self._invitations_controller.clear_invitations_for_operator(client_chat_id)
+            yield res
+
+    @contextmanager
+    def end_conversation_or_cancel_request_with_plocking(self,
+                                                         client_chat_id: int) -> Generator[Optional[int], None, None]:
+        with self._conversations_controller.end_conversation_or_cancel_request_with_plocking(client_chat_id) as \
+                operator_chat_id:
+
+            if operator_chat_id == -1:
+                self._invitations_controller.clear_invitations_to_client(client_chat_id)
+            elif operator_chat_id is not None:
+                self._invitations_controller.invite_for_operator(operator_chat_id)
+                # If this conversation's client is an operator, restore invitations for him, too
+                if self._users_controller.is_operator(client_chat_id):
```

Comment:
Add a note that there is no attempt to synchronize with operators list because the operators list is expected to not change while the application is running